File: workers/automation/src/jobctl/domain/pipeline/state_machine.py
"""StageStateMachine — pure function implementing the 16-transition table (ddd-target.md S8.5).

No I/O, no DB, no event publishing. Given a current StageState and a
StageTransition trigger, returns the new StageState or a TransitionRejected error.

Callers decide what to do with the result (publish events, persist, etc.).
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto
from typing import Callable

from jobctl.domain.pipeline_types import (
    Blocked,
    Canceled,
    Exhausted,
    Failed,
    Pending,
    Queued,
    Running,
    Skipped,
    Stale,
    StageState,
    Succeeded,
)


class StageTransition(Enum):
    """Triggers that drive state machine transitions."""

    Enqueue = auto()
    Start = auto()
    Complete = auto()
    Fail = auto()
    Block = auto()
    Skip = auto()
    Reset = auto()
    Cancel = auto()
    Exhaust = auto()
    Unblock = auto()
    MarkStale = auto()


@dataclass(frozen=True)
class TransitionRejected:
    """Returned when a transition is not allowed from the current state."""

    current_state: StageState
    attempted_transition: StageTransition
    reason: str


TransitionResult = StageState | TransitionRejected


def transition(current: StageState, trigger: StageTransition, **kwargs) -> TransitionResult:
    """Apply a state transition.  Pure function — no side effects.

    Keyword arguments are forwarded to the resulting StageState constructor
    (e.g. ``error_code``, ``reason``, ``blocked_by``).

    Returns the new StageState on success, or TransitionRejected on failure.
    """
    handler = _HANDLERS.get((current.kind, trigger))
    if handler is None:
        return TransitionRejected(
            current_state=current,
            attempted_transition=trigger,
            reason=f"Transition {trigger.name} is not allowed from {current.kind}",
        )
    return handler(current, **kwargs)


# ---------------------------------------------------------------------------
# Transition handlers — one per row of the S8.5 table
# ---------------------------------------------------------------------------


def _pending_to_queued(current: Pending, **kwargs) -> Queued:
    return Queued(queued_at=kwargs.get("queued_at", ""))


def _pending_to_running(current: Pending, **kwargs) -> Running:
    return Running(
        attempt_count=current.attempt_count + 1,
        started_at=kwargs.get("started_at", ""),
    )


def _pending_to_blocked(current: Pending, **kwargs) -> Blocked:
    return Blocked(
        blocked_by=kwargs.get("blocked_by", ()),
        error_code=kwargs.get("error_code", "BLOCKED_UPSTREAM"),
        error_message=kwargs.get("error_message", ""),
    )


def _pending_to_skipped(current: Pending, **kwargs) -> Skipped:
    return Skipped(reason=kwargs.get("reason", ""))


def _queued_to_running(current: Queued, **kwargs) -> Running:
    return Running(
        attempt_count=kwargs.get("attempt_count", 1),
        started_at=kwargs.get("started_at", ""),
    )


def _queued_to_canceled(current: Queued, **kwargs) -> Canceled:
    return Canceled(
        canceled_at=kwargs.get("canceled_at", ""),
        reason=kwargs.get("reason", None),
    )


def _running_to_succeeded(current: Running, **kwargs) -> Succeeded:
    return Succeeded(
        attempt_count=current.attempt_count,
        finished_at=kwargs.get("finished_at", ""),
        duration_ms=kwargs.get("duration_ms", 0),
    )


def _running_to_failed(current: Running, **kwargs) -> Failed:
    return Failed(
        attempt_count=current.attempt_count,
        max_attempts=kwargs.get("max_attempts", 0),
        error_code=kwargs.get("error_code", ""),
        error_message=kwargs.get("error_message", ""),
        retryable=kwargs.get("retryable", True),
        next_action=kwargs.get("next_action", None),
    )


def _running_to_canceled(current: Running, **kwargs) -> Canceled:
    return Canceled(
        canceled_at=kwargs.get("canceled_at", ""),
        reason=kwargs.get("reason", None),
    )


def _failed_to_pending(current: Failed, **kwargs) -> Pending:
    reset_attempts = kwargs.get("reset_attempts", False)
    return Pending(
        attempt_count=0 if reset_attempts else current.attempt_count,
        max_attempts=current.max_attempts,
        next_action=kwargs.get("next_action", None),
    )


def _failed_to_exhausted(current: Failed, **kwargs) -> Exhausted:
    return Exhausted(
        attempt_count=current.attempt_count,
        max_attempts=current.max_attempts,
        error_code=kwargs.get("error_code", current.error_code),
        error_message=kwargs.get("error_message", current.error_message),
        next_action=kwargs.get("next_action", None),
    )


def _blocked_to_pending(current: Blocked, **kwargs) -> Pending:
    return Pending(
        attempt_count=kwargs.get("attempt_count", 0),
        max_attempts=kwargs.get("max_attempts", 0),
        next_action=kwargs.get("next_action", None),
    )


def _exhausted_to_pending(current: Exhausted, **kwargs) -> Pending:
    return Pending(
        attempt_count=0,
        max_attempts=current.max_attempts,
        next_action=kwargs.get("next_action", None),
    )


def _canceled_to_pending(current: Canceled, **kwargs) -> Pending:
    return Pending(
        attempt_count=kwargs.get("attempt_count", 0),
        max_attempts=kwargs.get("max_attempts", 0),
        next_action=kwargs.get("next_action", None),
    )


def _succeeded_to_stale(current: Succeeded, **kwargs) -> Stale:
    return Stale(reason=kwargs.get("reason", ""))


def _stale_to_pending(current: Stale, **kwargs) -> Pending:
    return Pending(
        attempt_count=kwargs.get("attempt_count", 0),
        max_attempts=kwargs.get("max_attempts", 0),
        next_action=kwargs.get("next_action", None),
    )


# Dispatch table keyed by (current_kind, trigger). Each handler is an
# ``(StageState, **kwargs) -> StageState`` callable; the type ignore on the
# row signatures is acceptable here because the dispatch is value-based and
# the handlers themselves enforce the input ``StageState`` subclass at call
# time via positional typing.
_Handler = Callable[..., StageState]
_HANDLERS: dict[tuple[str, StageTransition], _Handler] = {
    # Row 1:  Pending  -> Queued   (Enqueue)
    ("Pending", StageTransition.Enqueue): _pending_to_queued,
    # Row 2:  Pending  -> Running  (Start)
    ("Pending", StageTransition.Start): _pending_to_running,
    # Row 3:  Pending  -> Blocked  (Block)
    ("Pending", StageTransition.Block): _pending_to_blocked,
    # Row 4:  Pending  -> Skipped  (Skip)
    ("Pending", StageTransition.Skip): _pending_to_skipped,
    # Row 5:  Queued   -> Running  (Start)
    ("Queued", StageTransition.Start): _queued_to_running,
    # Row 6:  Queued   -> Canceled (Cancel)
    ("Queued", StageTransition.Cancel): _queued_to_canceled,
    # Row 7:  Running  -> Succeeded (Complete)
    ("Running", StageTransition.Complete): _running_to_succeeded,
    # Row 8:  Running  -> Failed   (Fail)
    ("Running", StageTransition.Fail): _running_to_failed,
    # Row 9:  Running  -> Canceled (Cancel)
    ("Running", StageTransition.Cancel): _running_to_canceled,
    # Row 10: Failed   -> Pending  (Reset)
    ("Failed", StageTransition.Reset): _failed_to_pending,
    # Row 11: Failed   -> Exhausted (Exhaust)
    ("Failed", StageTransition.Exhaust): _failed_to_exhausted,
    # Row 12: Blocked  -> Pending  (Unblock)
    ("Blocked", StageTransition.Unblock): _blocked_to_pending,
    # Row 13: Exhausted -> Pending (Reset)
    ("Exhausted", StageTransition.Reset): _exhausted_to_pending,
    # Row 14: Canceled -> Pending  (Reset)
    ("Canceled", StageTransition.Reset): _canceled_to_pending,
    # Row 15: Succeeded -> Stale   (MarkStale)
    ("Succeeded", StageTransition.MarkStale): _succeeded_to_stale,
    # Row 16: Stale    -> Pending  (Reset)
    ("Stale", StageTransition.Reset): _stale_to_pending,
}


# ---------------------------------------------------------------------------
# Valid (from_kind, to_kind) pairs — derived from handler return type annotations.
# Used by set_stage_state for transition validation without needing a trigger.
# ---------------------------------------------------------------------------

_VALID_KIND_TRANSITIONS: frozenset[tuple[str, str]] = frozenset(
    {
        ("Pending", "Queued"),
        ("Pending", "Running"),
        ("Pending", "Blocked"),
        ("Pending", "Skipped"),
        ("Queued", "Running"),
        ("Queued", "Canceled"),
        ("Running", "Succeeded"),
        ("Running", "Failed"),
        ("Running", "Canceled"),
        ("Failed", "Pending"),
        ("Failed", "Exhausted"),
        ("Blocked", "Pending"),
        ("Exhausted", "Pending"),
        ("Canceled", "Pending"),
        ("Succeeded", "Stale"),
        ("Stale", "Pending"),
    }
)


def is_valid_transition(from_kind: str, to_kind: str) -> bool:
    """Check if transitioning from *from_kind* to *to_kind* is allowed.

    Uses the §8.5 transition table. Both arguments are PascalCase kind names.
    This is a quick lookup for callers (like ``set_stage_state``) that know the
    target state but not the trigger.
    """
    return (from_kind, to_kind) in _VALID_KIND_TRANSITIONS

File: packages/domain-types/src/pipeline/state_machine.ts
/**
 * StageStateMachine — TS port of `workers/automation/src/jobctl/domain/pipeline/state_machine.py`.
 *
 * Pure functions, no I/O, no runtime dependencies. Parity with the Python
 * implementation is enforced by `test/state_machine_parity.test.ts` plus the
 * sibling `tests/test_state_machine_parity.py`, both driven from the shared
 * fixture in `test/fixtures/state_machine_transitions.json`.
 *
 * Mirrors §8.5 — sixteen valid (fromKind, toKind) pairs.
 */

import type {
  Blocked,
  Canceled,
  Exhausted,
  Failed,
  Pending,
  Queued,
  Running,
  Skipped,
  Stale,
  StageState,
  StageStateKind,
  Succeeded,
} from "../pipeline.js";

/* ------------------------------------------------------------------ trigger */

/** Triggers that drive state machine transitions (parity with Python enum). */
export const StageTransitions = [
  "Enqueue",
  "Start",
  "Complete",
  "Fail",
  "Block",
  "Skip",
  "Reset",
  "Cancel",
  "Exhaust",
  "Unblock",
  "MarkStale",
] as const;
export type StageTransition = (typeof StageTransitions)[number];

/* ----------------------------------------------------------------- rejected */

export interface TransitionRejected {
  readonly kind: "TransitionRejected";
  readonly currentState: StageState;
  readonly attemptedTransition: StageTransition;
  readonly reason: string;
}

export type TransitionResult = StageState | TransitionRejected;

export function isRejected(value: TransitionResult): value is TransitionRejected {
  return (value as TransitionRejected).kind === "TransitionRejected";
}

/* --------------------------------------------------------- §8.5 valid pairs */

/** Sixteen valid (fromKind, toKind) transitions from §8.5. */
export const VALID_KIND_TRANSITIONS: ReadonlyArray<readonly [StageStateKind, StageStateKind]> = [
  ["Pending", "Queued"],
  ["Pending", "Running"],
  ["Pending", "Blocked"],
  ["Pending", "Skipped"],
  ["Queued", "Running"],
  ["Queued", "Canceled"],
  ["Running", "Succeeded"],
  ["Running", "Failed"],
  ["Running", "Canceled"],
  ["Failed", "Pending"],
  ["Failed", "Exhausted"],
  ["Blocked", "Pending"],
  ["Exhausted", "Pending"],
  ["Canceled", "Pending"],
  ["Succeeded", "Stale"],
  ["Stale", "Pending"],
] as const;

const VALID_KIND_TRANSITION_SET: ReadonlySet<string> = new Set(
  VALID_KIND_TRANSITIONS.map(([from, to]) => `${from}->${to}`),
);

/** Pure check: does the §8.5 table allow `fromKind → toKind`? */
export function isValidTransition(fromKind: StageStateKind, toKind: StageStateKind): boolean {
  return VALID_KIND_TRANSITION_SET.has(`${fromKind}->${toKind}`);
}

/* ----------------------------------------------------- transition handlers */

/**
 * Optional metadata you can pass through `applyTransition`. Mirrors the
 * **kwargs forwarded to the Python handlers — every field is optional.
 */
export interface TransitionInputs {
  readonly queuedAt?: string;
  readonly startedAt?: string;
  readonly finishedAt?: string;
  readonly canceledAt?: string;
  readonly attemptCount?: number;
  readonly maxAttempts?: number;
  readonly durationMs?: number;
  readonly errorCode?: string;
  readonly errorMessage?: string;
  readonly retryable?: boolean;
  readonly nextAction?: string;
  readonly reason?: string;
  readonly resetAttempts?: boolean;
  readonly blockedBy?: readonly string[];
}

type Handler = (current: StageState, inputs: TransitionInputs) => StageState;

const HANDLERS = new Map<string, Handler>();

function key(fromKind: StageStateKind, trigger: StageTransition): string {
  return `${fromKind}->${trigger}`;
}

function register(fromKind: StageStateKind, trigger: StageTransition, fn: Handler): void {
  HANDLERS.set(key(fromKind, trigger), fn);
}

// Row 1: Pending -> Queued
register("Pending", "Enqueue", (_current, inputs): Queued => ({
  kind: "Queued",
  queuedAt: inputs.queuedAt ?? "",
}));

// Row 2: Pending -> Running
register("Pending", "Start", (current, inputs): Running => {
  const pending = current as Pending;
  return {
    kind: "Running",
    attemptCount: pending.attemptCount + 1,
    startedAt: inputs.startedAt ?? "",
  };
});

// Row 3: Pending -> Blocked
register("Pending", "Block", (_current, inputs): Blocked => ({
  kind: "Blocked",
  blockedBy: (inputs.blockedBy ?? []) as Blocked["blockedBy"],
  errorCode: inputs.errorCode ?? "BLOCKED_UPSTREAM",
  errorMessage: inputs.errorMessage ?? "",
}));

// Row 4: Pending -> Skipped
register("Pending", "Skip", (_current, inputs): Skipped => ({
  kind: "Skipped",
  reason: inputs.reason ?? "",
}));

// Row 5: Queued -> Running
register("Queued", "Start", (_current, inputs): Running => ({
  kind: "Running",
  attemptCount: inputs.attemptCount ?? 1,
  startedAt: inputs.startedAt ?? "",
}));

// Row 6: Queued -> Canceled
register("Queued", "Cancel", (_current, inputs): Canceled => ({
  kind: "Canceled",
  canceledAt: inputs.canceledAt ?? "",
  reason: inputs.reason,
}));

// Row 7: Running -> Succeeded
register("Running", "Complete", (current, inputs): Succeeded => {
  const running = current as Running;
  return {
    kind: "Succeeded",
    attemptCount: running.attemptCount,
    finishedAt: inputs.finishedAt ?? "",
    durationMs: inputs.durationMs ?? 0,
  };
});

// Row 8: Running -> Failed
register("Running", "Fail", (current, inputs): Failed => {
  const running = current as Running;
  return {
    kind: "Failed",
    attemptCount: running.attemptCount,
    maxAttempts: inputs.maxAttempts ?? 0,
    errorCode: inputs.errorCode ?? "",
    errorMessage: inputs.errorMessage ?? "",
    retryable: inputs.retryable ?? true,
    nextAction: inputs.nextAction,
  };
});

// Row 9: Running -> Canceled
register("Running", "Cancel", (_current, inputs): Canceled => ({
  kind: "Canceled",
  canceledAt: inputs.canceledAt ?? "",
  reason: inputs.reason,
}));

// Row 10: Failed -> Pending
register("Failed", "Reset", (current, inputs): Pending => {
  const failed = current as Failed;
  const reset = inputs.resetAttempts ?? false;
  return {
    kind: "Pending",
    attemptCount: reset ? 0 : failed.attemptCount,
    maxAttempts: failed.maxAttempts,
    nextAction: inputs.nextAction,
  };
});

// Row 11: Failed -> Exhausted
register("Failed", "Exhaust", (current, inputs): Exhausted => {
  const failed = current as Failed;
  return {
    kind: "Exhausted",
    attemptCount: failed.attemptCount,
    maxAttempts: failed.maxAttempts,
    errorCode: inputs.errorCode ?? failed.errorCode,
    errorMessage: inputs.errorMessage ?? failed.errorMessage,
    nextAction: inputs.nextAction,
  };
});

// Row 12: Blocked -> Pending
register("Blocked", "Unblock", (_current, inputs): Pending => ({
  kind: "Pending",
  attemptCount: inputs.attemptCount ?? 0,
  maxAttempts: inputs.maxAttempts ?? 0,
  nextAction: inputs.nextAction,
}));

// Row 13: Exhausted -> Pending
register("Exhausted", "Reset", (current, inputs): Pending => {
  const exhausted = current as Exhausted;
  return {
    kind: "Pending",
    attemptCount: 0,
    maxAttempts: exhausted.maxAttempts,
    nextAction: inputs.nextAction,
  };
});

// Row 14: Canceled -> Pending
register("Canceled", "Reset", (_current, inputs): Pending => ({
  kind: "Pending",
  attemptCount: inputs.attemptCount ?? 0,
  maxAttempts: inputs.maxAttempts ?? 0,
  nextAction: inputs.nextAction,
}));

// Row 15: Succeeded -> Stale
register("Succeeded", "MarkStale", (_current, inputs): Stale => ({
  kind: "Stale",
  reason: inputs.reason ?? "",
}));

// Row 16: Stale -> Pending
register("Stale", "Reset", (_current, inputs): Pending => ({
  kind: "Pending",
  attemptCount: inputs.attemptCount ?? 0,
  maxAttempts: inputs.maxAttempts ?? 0,
  nextAction: inputs.nextAction,
}));

/* -------------------------------------------------------- applyTransition */

/**
 * Apply a state transition. Pure function — no side effects.
 *
 * Returns the new `StageState` on success or `TransitionRejected` when the
 * trigger is not allowed from `current.kind` per the §8.5 table.
 */
export function applyTransition(
  current: StageState,
  trigger: StageTransition,
  inputs: TransitionInputs = {},
): TransitionResult {
  const handler = HANDLERS.get(key(current.kind, trigger));
  if (!handler) {
    return {
      kind: "TransitionRejected",
      currentState: current,
      attemptedTransition: trigger,
      reason: `Transition ${trigger} is not allowed from ${current.kind}`,
    };
  }
  return handler(current, inputs);
}
